File: 0150-evaluate-reverse-polish-notation.py
"""
150. Evalute Reverse Polish Notation

Submitted: November 27, 2024

Runtime: 2 ms (beats 72.29%)
Memory: 19.11 MB (beats 6.55%)
"""

class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        stack = []
        for token in tokens:
            if token == '+':
                x = stack.pop()
                y = stack.pop()
                stack.append(x + y)
            elif token == "-": 
                y = stack.pop()
                x = stack.pop()
                stack.append(x - y)
            elif token == "*": 
                x = stack.pop()
                y = stack.pop()
                stack.append(x * y)
            elif token == "/": 
                y = stack.pop()
                x = stack.pop()
                stack.append(int(x / y))
            else:
                stack.append(int(token))
        return stack[-1]

File: 0150-evaluate-reverse-polish-notation.cpp
/*
150. Evalute Reverse Polish Notation

Submitted: November 16, 2024

Runtime: 0 ms (beats 100.00%)
Memory: 15.74 MB (beats 61.18%)
*/

class Solution {
public:
    int evalRPN(vector<string>& tokens) {
        stack<int> stack;
        for (const string& token : tokens) {
            if (token == "+") {
                int x = stack.top();
                stack.pop();
                int y = stack.top();
                stack.pop();
                stack.push(x + y);
            } else if (token == "-") {
                int y = stack.top();
                stack.pop();
                int x = stack.top();
                stack.pop();
                stack.push(x - y);
            } else if (token == "*") {
                int x = stack.top();
                stack.pop();
                int y = stack.top();
                stack.pop();
                stack.push(x * y);
            } else if (token == "/") {
                int y = stack.top();
                stack.pop();
                int x = stack.top();
                stack.pop();
                stack.push(x / y);
            } else {
                stack.push(stoi(token));
            }
        }
        return stack.top();
    }
};
